### ISSUE TITLE
Error due to long link line: Refactor makefile so we can link more object files

### ISSUE THREAD

**Issue by manauref** [ammarhakim/gkylzero#398]
In some branches of the code (e.g. gk-g0-app_vp_cudss) compiling the code is resulting in this error after `make install`:

```
...
.cu.o cuda-build/zero/bc_twistshift_cu.cu.o cuda-build/zero/fem_poisson_perp_cu.cu.o cuda-build/zero/bc_basic_cu.cu.o cuda-build/zero/cart_modal_serendip_cu.cu.o cuda-build/zero/dg_diffusion_gyrokinetic_cu.cu.o cuda-build/zero/cart_modal_gkhybrid_cu.cu.o cuda-build/zero/dg_calc_vlasov_gen_geo_vars_cu.cu.o -L/scratch/gpfs/manaurer/gkeyll/code/g0-vp/gkylsoft/superlu/lib -L/scratch/gpfs/manaurer/gkeyll/code/g0-vp/gkylsoft/OpenBLAS/lib -Lcuda-build -L/usr/local/openmpi/cuda-11.1/4.1.1/gcc/lib64 -L/usr/lib64 -L/scratch/gpfs/manaurer/gkeyll/code/g0-vp/gkylsoft/cuDSS/lib/ -L. -lsuperlu /scratch/gpfs/manaurer/gkeyll/code/g0-vp/gkylsoft/OpenBLAS/lib/libopenblas.a  -lcublas -lcusparse -lcusolver -Xlinker "-rpath,/usr/local/openmpi/cuda-11.1/4.1.1/gcc/lib64" -lmpi -lnccl -Xlinker "-rpath,/scratch/gpfs/manaurer/gkeyll/code/g0-vp/gkylsoft/cuDSS/lib/" -lcudss   -lm -lpthread -ldl -o cuda-build/libgkylzero-install.so
make: execvp: /bin/sh: Argument list too long
make: *** [Makefile:364: cuda-build/libgkylzero.so] Error 127
make: *** Waiting for unfinished jobs....
make: *** [Makefile:374: cuda-build/libgkylzero-install.so] Error 127
```

This is on stellar-amd, building with nvcc 12.4.

One could actually reroute the output of `make install` to a file, and then manually copy that link command (~130k characters) into the terminal without a problem. So the issue is not the shell, but that the `make` command cannot execute such long line. Potential solutions include:
1. Refactor the makefile so we have several smaller linking steps.
2. One can, instead of having `make` execute this command, have the make file `echo` it onto the terminal (see https://stackoverflow.com/questions/64201512/argument-list-too-long-when-linking-with-gnu-make/64202705#64202705).

Option 1 seems preferable.

**Comment by manauref** [comment]
This now happening in the gk-g0-app_clopen branch as well, preventing us from implementing the ability to do clopen calculations

**Comment by ammarhakim** [comment]
We need to fix this for sure. 

However, I would like to recommend that people not merely file a report and the let others figure out a fix. It is not hard to try a few things out oneself  and see if a quick fix can be developed. Though of course I can do this it is better if others help in these types of issues also. I can't be always relied to get to these issues quickly. Realize my time is spread across the project and in co-managing a large department. Hence, we all need to develop the skills to debug build & other issues and learn to fix them.

For example, @manauref you pointed out a solution in the Stack overflow question. Was this tried? Can you mod the Makefile to write the args to a file and mod the link command to read the args from a file?

Let us try a few things.

**Comment by ammarhakim** [comment]
I took a first look at this. Unfortunately the code in the clopen  is broken, and it has nothing to do with this issue. The regression tests are not compiling, even on a CPU. Let us fix that first (and ensure "make unit regression" runs to completion on CPUs) so I can help look into this more carefully.

**Comment by manauref** [comment]
The regression test rt_gk_d3d_iwl_3x2v_p1.c is not ready, so it shouldn't
be run or even compiled in this branch. That said, I fixed some typos in it
so it can compile.
The test we are using in this branch so far is unit/ctest_bc_twistshift.

**Comment by ammarhakim** [comment]
I understand if a regression test is not ready. But if one wants someone else to look at an issue, one should ensure the code compiles. It’s frustrating if the very first thing one does on checking out the branch, i.e make unit regression, fails. One’s motivation to try a fix goes away.In general it’s ok to have broken code in ones branch. But not, in general, if someone else is requested to help fix an issue. The zeroth order etiquette is to ensure code compiles.Sent from my iPhoneOn Aug 2, 2024, at 2:20 AM, Manaure Francisquez ***@***.***> wrote:﻿
The regression test rt_gk_d3d_iwl_3x2v_p1.c is not ready, so it shouldn't

be run or even compiled in this branch. That said, I fixed some typos in it

so it can compile.

The test we are using in this branch so far is unit/ctest_bc_twistshift.

**Comment by JunoRavin** [comment]
This issue was addressed with https://github.com/ammarhakim/gkylzero/pull/440 by renaming build->bld and cuda-build->cubld to give ourselves more breathing room in the final link line with the names of directories. 

Long term we will likely want to break up the build into separate shared libraries for the different major apps (moments, Vlasov, and gyrokinetics) to provide users the options to build only a subset of the system (e.g., a user who only wants to do fluid simulations does not need to spend time compiling our plethora of kernels for our DG equation systems). But for now I am closing this issue as complete since we have staved off the error message by reducing the number of characters in the name of the build directories by a factor of 2 (so we can add twice as many kernels as we currently have before we run into issues again).
